lighthouse-core/computed/bundle-analysis.js

```diff
@@ -0,0 +1,151 @@
+ * @property {LH.Artifacts.RawSourceMap} rawMap
+ * @property {LH.Artifacts.ScriptElement} script
+ * @property {LH.Artifacts.NetworkRequest=} networkRecord
+ * @property {TextSourceMap} map
+ * @property {Sizes} sizes
+*/
+
+/**
+ * Calculate the number of bytes contributed by each source file
+ * @param {TextSourceMap} map
+ * @param {string} content
+ */
+function computeGeneratedFileSizes(map, content) {
+  const lines = content.split('\n');
+  /** @type {Record<string, number>} */
+  const files = {};
+  const totalBytes = content.length;
+  let unmappedBytes = totalBytes;
+
+  // If the map + contents don't line up, return a result that
+  // denotes nothing is mapped.
+  const failureResult = {files: {}, unmappedBytes, totalBytes};
+
+  // @ts-ignore: This function is added in SDK.js
+  map.computeLastGeneratedColumns();
+
+  for (const mapping of map.mappings()) {
+    const source = mapping.sourceURL;
+    const lineNum = mapping.lineNumber;
+    const colNum = mapping.columnNumber;
+    // @ts-ignore: `lastColumnNumber` is not on types yet.
+    const lastColNum = /** @type {number=} */ (mapping.lastColumnNumber);
+
+    // Webpack sometimes emits null mappings.
+    // https://github.com/mozilla/source-map/pull/303
+    if (!source) continue;
+
+    // Lines are 1-based
+    const line = lines[lineNum];
+    if (line === null) {
+      log.error(`${map.compiledURL} mapping for line out of bounds: ${lineNum + 1}`);
+      return failureResult;
+    }
+
+    // Columns are 0-based
+    if (colNum >= line.length) {
+      log.error(`${map.compiledURL} mapping for column out of bounds: ${lineNum + 1}:${colNum}`);
+      return failureResult;
+    }
+
+    let mappingLength = 0;
+    if (lastColNum !== undefined) {
+      if (lastColNum >= line.length) {
+        // eslint-disable-next-line max-len
+        log.error(`${map.compiledURL} mapping for last column out of bounds: ${lineNum + 1}:${lastColNum}`);
+        return failureResult;
+      }
+      mappingLength = lastColNum - colNum;
+    } else {
+      // TODO Buffer.byteLength?
+      mappingLength = line.length - colNum;
+    }
+    files[source] = (files[source] || 0) + mappingLength;
+    unmappedBytes -= mappingLength;
+  }
+
+  return {
+    files,
+    unmappedBytes,
+    totalBytes,
+  };
+}
+
+class BundleAnalysis {
+  /**
+   * @param {Pick<LH.Artifacts, 'SourceMaps'|'ScriptElements'|'devtoolsLogs'>} artifacts
+   * @param {LH.Audit.Context} context
+   */
+  static async compute_(artifacts, context) {
+    const {SourceMaps, ScriptElements} = artifacts;
+    const devtoolsLog = artifacts.devtoolsLogs[Audit.DEFAULT_PASS];
+    const networkRecords = await NetworkRecords.request(devtoolsLog, context);
+
+    /** @type {Bundle[]} */
+    const bundles = [];
+
+    // Collate map, script, and network record.
+    for (let mapIndex = 0; mapIndex < SourceMaps.length; mapIndex++) {
+      const SourceMap = SourceMaps[mapIndex];
+      if (!SourceMap.map) continue;
```

Comment:
is this possible?

Reply:
A source map declaration resulting in an error (fetch, bad json, etc.) still generates a value in `artifacts.SourceMaps`. Either `error` or `map` is set.